_posts/2022-11-12-Aya-gsoc-final-report.md

```diff
@@ -0,0 +1,132 @@
+
+- checks if the Guesser class support guessing the attributes passed to the`to_guess` and/or `force_guess` parameters.
+
+- check if any attribute passed to the `to_guess` parameter already exists in the topology attributes of the [Universe]. If so, warn the user that only empty values will be filled for this attribute, if any exists and in case the user wishes to override all the attribute values, he must pass it to the `force_guess` parameter instead of the `to_guess` one.
+
+- guessing attributes is handled by the guess_attrs method, which is declared inside the parent guesser class GuesserBase. It manages partial and complete guessing of attributes and calls the appropriate guesser method for each attribute.
+
+- each attribute guesser method searches for the reference attribute to begin guessing from it, and if not found in the [Universe], it calls the `guess_topologyAttributes`  to try guessing this reference attribute.
+
+- after guessing the attribute, the API adds it to the [Universe] with the help of `add_topologyAttr` or `_add_topology_objects` [Universe]’s methods
+
+Example of using the `guess_topologyAttributes` at [Universe] initiation:
+
+```python
+# to guess bonds for a [Universe]:
+ 
+     import MDAnalysis as mda
+     from MDAnalysisTests.datafiles import two_water_gro
+ 
+     u = mda.Universe(two_water_gro, context = 'default', to_guess=['bonds'])
+```
+
+Example of using the `guess_topologyAttributes` directly:
+
+```python
+# guess masses and types attribute::
+ 
+u.guess_TopologyAttributes(context='default', to_guess=['masses', 'types'])
+```
+More explanation is found in the user guide here: [guess_topologyAttributes], [Guessing]
+
+3. DefaultGuesser class
+The `DefaultGuesser` class holds the same old guessing methods but with modifying them to be compatible with the new guess_topologyAttrsibutes API. Moreover, I added a new feature to type guessing method so that it now can guess types from masses if names are not available ([commit 348f62d])
+
+4. Testing the old parser behavior is preserved
+The `to_guess` parameters of the [Universe] have a default value of (`types`, `masses`) to maintain the default behavior of the parsers. To make sure of not breaking old behavior, I added three tests in the parser's base.py module to check three things: 1- types and masses are guessed as expected in all [Universe]s created with the parsers after removing guessing from them 2- The values of the guessed types with the `guess_topologyAttributes` API are the same as those from the old behavior. 3- The values of the guessed masses with the `guess_topologyAttributes` API are the same as those from the old behavior. Once those three tests are passed, then it's safe to say that we are not breaking the default behavior of the code ([commit af84927]).
+
+I also discovered a bug in Topology's methods gussed_attributes and read_attributes  while working on developing the guesser API and fixed it ([merged PR # 3779])
```

Comment:
I also discovered a bug in the `guessed_attributes()` and `read_attributes()` methods of the `Topology` class ...